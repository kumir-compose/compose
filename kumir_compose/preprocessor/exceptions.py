from kumir_compose.preprocessor.tokens import DIRECTIVES


class LexerException(Exception):
    """Base class for lexer exceptions."""

    def __init__(
            self,
            message: str,
            filename: str,
            source: str,
            line: int,
            char: int
    ) -> None:
        """Create exception."""
        self.message = message
        self.filename = filename
        self.src = source
        self.line = line
        self.char = char

    @property
    def formatted_message(self) -> str:
        """Create beautiful representation of exception."""
        line_str = self.src.splitlines()[self.line - 1]
        line_str_prefix = f"{self.line} |  "
        pointer_margin = " " * (len(line_str_prefix) + self.char)
        return (
            f"{self.message}\n"
            f"At file {self.filename}, line {self.line}, char {self.char}:\n"
            f"{line_str_prefix}{line_str}\n"
            f"{pointer_margin}^"
        )

    def __str__(self):
        """Convert to string repr."""
        return self.formatted_message


class UnexpectedCharacterException(LexerException):
    """Raised when encountered unexpected character."""

    def __init__(
            self,
            got: str,
            expected: str,
            filename: str,
            source: str,
            line: int,
            char: int
    ) -> None:
        """Create exception and message."""
        got = f"'{got}'" if got else "END OF FILE"
        super().__init__(
            f"Unexpected character {got}, expected '{expected}'",
            filename,
            source,
            line,
            char,
        )


class UnknownDirectiveException(LexerException):
    """Raised when encountered unexpected directive."""

    def __init__(
            self,
            got: str,
            filename: str,
            source: str,
            line: int,
            char: int
    ) -> None:
        """Create exception and message."""
        super().__init__(
            (
                f"Unexpected character '{got}', "
                f"expected one of {DIRECTIVES.keys()}"
            ),
            filename,
            source,
            line,
            char,
        )
