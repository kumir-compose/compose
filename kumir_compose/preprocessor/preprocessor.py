from collections.abc import Collection, Sequence
from pathlib import Path

from attrs import frozen

from kumir_compose.preprocessor.exceptions import (
    IncludeFileNotFoundException,
    UnexpectedTokenException,
)
from kumir_compose.preprocessor.lexer import scan_file
from kumir_compose.preprocessor.tokens import Token, TokenType


@frozen
class Condition:
    name: str
    expected: bool


class Preprocessor:
    def __init__(
            self,
            filename: str,
            source: str,
            tokens: Sequence[Token],
            *,
            lookup_paths: Collection[str],
            encoding: str | None = None
    ) -> None:
        self._tokens = list(tokens)
        self._pos = 0
        self._filename = filename
        self._source = source
        self._macro_table: dict[str, Sequence[Token]] = {}
        self._lookup_paths = set(lookup_paths)
        self._lookup_paths.add("")
        self._encoding = encoding
        self._condition_stack = []
        self._included = set()

    @property
    def _at_end(self) -> bool:
        return self._pos >= len(self._tokens)

    def _next(self) -> Token | None:
        if self._at_end:
            return None
        self._pos += 1
        return self._tokens[self._pos - 1]

    @property
    def _peek(self) -> Token | None:
        if self._at_end:
            return None
        return self._tokens[self._pos]

    @property
    def _previous(self) -> Token:
        return self._tokens[self._pos - 1]

    def _match(
            self,
            expected_type: TokenType,
    ) -> Token | None:
        if self._at_end:
            return None
        if self._peek.type == expected_type:
            return self._next()
        return None

    def _error(self, exc_type, offset: int = 0, *args, **kwargs):
        kwargs["source"] = self._source
        kwargs["filename"] = self._filename
        kwargs["token"] = self._tokens[
            min(len(self._tokens) - 1, max(self._pos + offset, 0))
        ]
        raise exc_type(*args, **kwargs)

    def _require(
            self,
            expected_type: TokenType,
            *,
            message: str | None = None,
    ) -> Token:
        token = self._match(expected_type)
        if token is None:
            self._error(
                UnexpectedTokenException,
                expected=expected_type,
                got=self._peek,
                message=message,
            )
        return token

    def _consume_spaces(self) -> None:
        while self._match(TokenType.SPACE): ...

    def process(self):
        """Run preprocessor."""
        processed = ""
        while not self._at_end:
            token = self._next()
            if token.type == TokenType.ENDIF:
                self._end_conditional()
                continue
            if not self._do_conditions_match():
                continue
            match token.type:
                case TokenType.DEFINE:
                    self._define_macro()
                case TokenType.UNDEF:
                    self._undef_macro()
                case TokenType.INCLUDE:
                    self._include()
                case TokenType.IFDEF:
                    self._start_conditional(expected=True)
                case TokenType.IFNDEF:
                    self._start_conditional(expected=False)
                case TokenType.ID:
                    new_token = self._process_id(token)
                    if new_token.lexeme != token.lexeme:
                        self._pos -= 1
                    else:
                        processed += new_token.lexeme
                case _:
                    processed += token.lexeme
        return processed

    def _define_macro(self) -> None:
        self._consume_spaces()
        macro_name = self._require(TokenType.ID)
        self._consume_spaces()
        macro_value = self._next()
        if macro_value.type == TokenType.NEWLINE:
            self._macro_table[macro_name.value] = []
            return
        macro_values = [macro_value]
        while self._peek and self._peek.type != TokenType.NEWLINE:
            macro_values.append(self._next())
        self._macro_table[macro_name.value] = macro_values

    def _undef_macro(self) -> None:
        self._consume_spaces()
        macro_name = self._require(TokenType.ID)
        if macro_name.value in self._macro_table:
            self._macro_table.pop(macro_name.value)

    def _include(self) -> None:
        self._consume_spaces()
        path_tok = self._match(TokenType.ID)
        if not path_tok:
            path_tok = self._require(TokenType.STRING)
        found_path = self._find_file(path_tok.value)
        if not found_path:
            self._error(
                IncludeFileNotFoundException,
                incl_file=path_tok.value,
                offset=-2
            )
        abs_path = str(found_path.absolute())
        if abs_path in self._included:
            return
        tokens = scan_file(abs_path, self._encoding)
        self._insert_tokens_here(list(tokens))
        self._included.add(abs_path)

    def _insert_tokens_here(self, tokens: list[Token]) -> None:
        self._tokens = (
            self._tokens[:self._pos] +
            list(tokens) +
            self._tokens[self._pos:]
        )

    def _find_file(self, filename) -> Path | None:
        for root in self._lookup_paths:
            if Path(root, filename).exists():
                return Path(root, filename)
            if Path(root, f"{filename}.kum").exists():
                return Path(root, f"{filename}.kum")
        return None

    def _start_conditional(self, *, expected: bool) -> None:
        self._consume_spaces()
        def_name = self._require(TokenType.ID)
        self._condition_stack.append(Condition(
            name=def_name.value,
            expected=expected
        ))

    def _end_conditional(self) -> None:
        self._condition_stack.pop()

    def _do_conditions_match(self) -> bool:
        for condition in self._condition_stack:
            is_present = condition.name in self._macro_table
            if is_present != condition.expected:
                return False
        return True

    def _process_id(self, token: Token) -> Token:
        if token.lexeme in self._macro_table:
            first_token, *other_tokens = self._macro_table[token.lexeme]
            self._tokens[self._pos - 1] = first_token
            self._insert_tokens_here(other_tokens)
            return first_token
        return token
