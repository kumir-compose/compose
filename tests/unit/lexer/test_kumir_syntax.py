from operator import attrgetter

from kumir_compose.preprocessor.lexer import Lexer
from tests.utils import read_fix_file


def test_basic_kumir_syntax(snapshot):
    """Test that Kumir syntax is parsed."""
    lexer = Lexer("", read_fix_file("all_possible_lexemes.kum"))
    tokens = lexer.scan()
    str_repr = "\n".join(map(str, tokens))
    assert str_repr == snapshot


def test_original_file_restorable():
    """Test that we can recreate original file from tokens."""
    orig_source = read_fix_file("all_possible_lexemes.kum")
    lexer = Lexer("", orig_source)
    tokens = lexer.scan()
    restored_source = "".join(map(attrgetter("lexeme"), tokens))
    orig_source = orig_source.replace("\r", "")
    assert restored_source == orig_source
