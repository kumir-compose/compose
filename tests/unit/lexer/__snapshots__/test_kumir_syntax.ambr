# serializer version: 1
# name: test_basic_kumir_syntax
  '''
  TokenType.KEYWORD:аргрез:1:6
  TokenType.SPACE: :1:7
  TokenType.KEYWORD:знач:1:11
  TokenType.SPACE: :1:12
  TokenType.KEYWORD:цел:1:15
  TokenType.SPACE: :1:16
  TokenType.KEYWORD:вещ:1:19
  TokenType.SPACE: :1:20
  TokenType.KEYWORD:лог:1:23
  TokenType.SPACE: :1:24
  TokenType.KEYWORD:сим:1:27
  TokenType.SPACE: :1:28
  TokenType.KEYWORD:лит:1:31
  TokenType.SPACE: :1:32
  TokenType.ID:таб:1:35
  TokenType.NEWLINE:
  :1:36
  TokenType.KEYWORD:целтаб:2:6
  TokenType.SPACE: :2:7
  TokenType.KEYWORD:вещтаб:2:13
  TokenType.SPACE: :2:14
  TokenType.KEYWORD:логтаб:2:20
  TokenType.SPACE: :2:21
  TokenType.KEYWORD:литтаб:2:27
  TokenType.SPACE: :2:28
  TokenType.KEYWORD:и:2:29
  TokenType.SPACE: :2:30
  TokenType.KEYWORD:или:2:33
  TokenType.SPACE: :2:34
  TokenType.ID:не:2:36
  TokenType.NEWLINE:
  :2:37
  TokenType.KEYWORD:да:3:2
  TokenType.SPACE: :3:3
  TokenType.KEYWORD:нет:3:6
  TokenType.SPACE: :3:7
  TokenType.KEYWORD:утв:3:10
  TokenType.SPACE: :3:11
  TokenType.KEYWORD:выход:3:16
  TokenType.SPACE: :3:17
  TokenType.KEYWORD:ввод:3:21
  TokenType.SPACE: :3:22
  TokenType.KEYWORD:вывод:3:27
  TokenType.SPACE: :3:28
  TokenType.KEYWORD:нс:3:30
  TokenType.SPACE: :3:31
  TokenType.ID:если:3:35
  TokenType.NEWLINE:
  :3:36
  TokenType.KEYWORD:то:4:2
  TokenType.SPACE: :4:3
  TokenType.KEYWORD:иначе:4:8
  TokenType.SPACE: :4:9
  TokenType.KEYWORD:все:4:12
  TokenType.SPACE: :4:13
  TokenType.KEYWORD:выбор:4:18
  TokenType.SPACE: :4:19
  TokenType.KEYWORD:при:4:22
  TokenType.SPACE: :4:23
  TokenType.KEYWORD:нц:4:25
  TokenType.SPACE: :4:26
  TokenType.KEYWORD:кц:4:28
  TokenType.SPACE: :4:29
  TokenType.ID:кц_при:4:35
  TokenType.NEWLINE:
  :4:36
  TokenType.KEYWORD:раз:5:3
  TokenType.SPACE: :5:4
  TokenType.KEYWORD:пока:5:8
  TokenType.SPACE: :5:9
  TokenType.KEYWORD:для:5:12
  TokenType.SPACE: :5:13
  TokenType.KEYWORD:от:5:15
  TokenType.SPACE: :5:16
  TokenType.KEYWORD:до:5:18
  TokenType.SPACE: :5:19
  TokenType.KEYWORD:шаг:5:22
  TokenType.SPACE: :5:23
  TokenType.KEYWORD:алг:5:26
  TokenType.SPACE: :5:27
  TokenType.KEYWORD:нач:5:30
  TokenType.SPACE: :5:31
  TokenType.ID:кон:5:34
  TokenType.NEWLINE:
  :5:35
  TokenType.KEYWORD:исп:6:3
  TokenType.SPACE: :6:4
  TokenType.KEYWORD:кон_исп:6:11
  TokenType.SPACE: :6:12
  TokenType.KEYWORD:дано:6:16
  TokenType.SPACE: :6:17
  TokenType.KEYWORD:надо:6:21
  TokenType.SPACE: :6:22
  TokenType.KEYWORD:арг:6:25
  TokenType.SPACE: :6:26
  TokenType.ID:рез:6:29
  TokenType.NEWLINE:
  :6:30
  TokenType.PUNCTUATION:(:7:1
  TokenType.PUNCTUATION:):7:2
  TokenType.PUNCTUATION:[:7:3
  TokenType.PUNCTUATION:]:7:4
  TokenType.PUNCTUATION:+:7:5
  TokenType.PUNCTUATION:-:7:6
  TokenType.PUNCTUATION:*:7:7
  TokenType.PUNCTUATION:/:7:8
  TokenType.PUNCTUATION:,:7:9
  TokenType.PUNCTUATION:<:7:10
  TokenType.PUNCTUATION:>:7:11
  TokenType.PUNCTUATION:::7:12
  TokenType.PUNCTUATION:=:7:13
  TokenType.NEWLINE:
  :7:14
  TokenType.KEYWORD:вещ:8:3
  TokenType.SPACE: :8:4
  TokenType.ID:number:8:11
  TokenType.SPACE: :8:12
  TokenType.PUNCTUATION:=:8:13
  TokenType.SPACE: :8:14
  TokenType.NUMBER:132.32332:8:23
  TokenType.NEWLINE:
  :8:24
  TokenType.KEYWORD:цел:9:3
  TokenType.SPACE: :9:4
  TokenType.ID:n:9:6
  TokenType.SPACE: :9:7
  TokenType.PUNCTUATION:=:9:8
  TokenType.SPACE: :9:9
  TokenType.NUMBER:123:9:12
  TokenType.NEWLINE:
  :9:13
  TokenType.KEYWORD:лит:10:3
  TokenType.SPACE: :10:4
  TokenType.ID:имя:10:8
  TokenType.SPACE: :10:9
  TokenType.PUNCTUATION:=:10:10
  TokenType.SPACE: :10:11
  TokenType.STRING:"это двойные кавычки":10:32
  TokenType.NEWLINE:
  :10:33
  TokenType.KEYWORD:лит:11:3
  TokenType.SPACE: :11:4
  TokenType.ID:имя с пробелом_ да_ так тоже можно:11:39
  TokenType.SPACE: :11:40
  TokenType.PUNCTUATION:=:11:41
  TokenType.SPACE: :11:42
  TokenType.STRING:'это одинарные кавычки':11:65
  TokenType.NEWLINE:
  :11:66
  TokenType.COMMENT:| комментарий:12:13
  TokenType.NEWLINE:
  :12:14
  TokenType.COMMENT:| еще комментарий:13:17
  TokenType.NEWLINE:
  :13:18
  TokenType.KEYWORD:лог:14:3
  TokenType.SPACE: :14:4
  TokenType.ID:ехать на дачу:14:18
  TokenType.SPACE: :14:19
  TokenType.PUNCTUATION:=:14:20
  TokenType.SPACE: :14:21
  TokenType.ID:сегодня солнечно:14:39
  TokenType.SPACE: :14:424
  TokenType.KEYWORD:и:14:39
  TokenType.SPACE: :14:40
  TokenType.ID:завтра солнечно:14:55
  TokenType.NEWLINE:
  :14:56
  '''
# ---
